torch/_dynamo/variables/user_defined.py

```diff
@@ -630,6 +631,41 @@ def deque_signature(iterable=None, maxlen=None):
                             "Set torch._dynamo.config.enable_trace_contextlib = True",
                         ],
                     )
+
+                from . import SDPAKernelVariable, UserFunctionVariable
+
+                # Special treatments for certain context managers created via
+                # contextlib, because
+                # 1. we (pytorch) own their impls
+                # 2. it's tedious to trace through them, so we effectively
+                #    "allow in graph" them without sacrificing soundness.
+                #
+                # We would typically reach here via either
+                # 1. the instance construction in `with ctx_manager(...):`:
+                #    https://github.com/python/cpython/blob/3.12/Lib/contextlib.py#L301
+                # 2. calling a function decorated with a context manager:
+                #    https://github.com/python/cpython/blob/3.12/Lib/contextlib.py#L122
+                #
+                # So we basically trace through the surface part of the
+                # contextlib code, and then special case the shared remaining
+                # logic (the actual context manager instance construction and
+                # usage later on).
+                if isinstance(args[0], UserFunctionVariable):
+                    fn = args[0].fn
+                    fn_source = args[0].source
+                    # At this point `fn` is the "unwrapped" version of the
+                    # context-managerified function.
+                    if fn is torch.nn.attention.sdpa_kernel.__wrapped__:
+                        args, kwargs = args[1], args[2]
+                        name_to_arg_map = bind_args_cached(
+                            fn, tx, fn_source, args.items, kwargs.items
+                        )
+                        backends = name_to_arg_map["backends"].as_python_constant()
+                        set_priority = name_to_arg_map[
+                            "set_priority"
+                        ].as_python_constant()
+                        return SDPAKernelVariable.create(tx, backends, set_priority)
```

Comment:
There's a `TorchCtxManagerClassVariable` that handles special cases of context managers. Could you try to implement this context manager using it?

https://github.com/pytorch/pytorch/blob/30d2f98daa956843a8f7b43e809da57910540306/torch/_dynamo/variables/torch.py#L13-L14

Reply:
I was basically moving the `sdpa_kernels` special casing out of it, since it's no longer a "class value" anymore after this patch (we'll trace into the actual wrapped `sdpa_kernels` as a function value), and I was intercepting the actual context manager construction right here.

Regardless, keeping the logic in `TorchCtxManagerClassVariable` might also make things more organized anyways, despite the misnomer. Ideally I'm hoping we just trace through the python code as much as possible eventually, without special casing, to avoid issues like #160684.